test/opt/test_fold_spec_const_op_composite.cpp

```diff
@@ -0,0 +1,1400 @@
+  builder.AppendTypesConstantsGlobals({
+      // clang-format off
+        "%int = OpTypeInt 32 1",
+        "%frozen_spec_const_int = OpConstant %int 1",
+        "%const_int = OpConstant %int 2",
+        // Folding target:
+        "%spec_add = OpSpecConstantOp %int IAdd %frozen_spec_const_int %const_int",
+      // clang-format on
+  });
+
+  std::vector<const char*> expected = {
+      // clang-format off
+                    "OpCapability Shader",
+                    "OpCapability Float64",
+               "%1 = OpExtInstImport \"GLSL.std.450\"",
+                    "OpMemoryModel Logical GLSL450",
+                    "OpEntryPoint Vertex %main \"main\"",
+                    "OpName %void \"void\"",
+                    "OpName %main_func_type \"main_func_type\"",
+                    "OpName %main \"main\"",
+                    "OpName %main_func_entry_block \"main_func_entry_block\"",
+                    "OpName %int \"int\"",
+                    "OpName %frozen_spec_const_int \"frozen_spec_const_int\"",
+                    "OpName %const_int \"const_int\"",
+                    "OpName %spec_add \"spec_add\"",
+            "%void = OpTypeVoid",
+  "%main_func_type = OpTypeFunction %void",
+             "%int = OpTypeInt 32 1",
+"%frozen_spec_const_int = OpConstant %int 1",
+       "%const_int = OpConstant %int 2",
+        // The SpecConstantOp IAdd instruction should be replace by OpConstant
+        // instruction:
+        "%spec_add = OpConstant %int 3",
+            "%main = OpFunction %void None %main_func_type",
+"%main_func_entry_block = OpLabel",
+                    "OpReturn",
+                    "OpFunctionEnd",
+      // clang-format on
+  };
+  SinglePassRunAndCheck<opt::FoldSpecConstantOpAndCompositePass>(
+      builder.GetCode(), JoinAllInsts(expected), /* skip_nop = */ true);
+}
+
+// A test of skipping folding an instruction when the instruction result type
+// has decorations.
+TEST_F(FoldSpecConstantOpAndCompositePassBasicTest,
+       SkipWhenTypeHasDecorations) {
+  AssemblyBuilder builder;
+  builder
+      .AppendAnnotations({
+          // clang-format off
+    "OpDecorate %int RelaxedPrecision",
+          // clang-format on
+      })
+      .AppendTypesConstantsGlobals({
+          // clang-format off
+        "%int = OpTypeInt 32 1",
+        "%frozen_spec_const_int = OpConstant %int 1",
+        "%const_int = OpConstant %int 2",
+        // The following spec constant should not be folded as the result type
+        // has relaxed precision decoration.
+        "%spec_add = OpSpecConstantOp %int IAdd %frozen_spec_const_int %const_int",
+          // clang-format on
+      });
+
+  SinglePassRunAndCheck<opt::FoldSpecConstantOpAndCompositePass>(
+      builder.GetCode(), builder.GetCode(), /* skip_nop = */ true);
+}
+
+// All types and some common constants that are potentially required in
+// FoldSpecConstantOpAndCompositeTest.
+const std::vector<std::string> CommonTypesAndConstants = {
```

Comment:
The style guide forbids non-POD class variables with static duration.
Make this a function that returns the vector.